![Marcel](pics/marcel.png) 

#  Segmenting and Clustering Neighborhoods in Toronto 


- __This is assigment from week 3  module 9 for the completion of the IBM Data Science Professional Certificate__

<HR>

> # __Assignment 1:__
>
>    For this assignment, I will explore and cluster the neighborhoods in Toronto.
>
>    1. I start by creating this Notebook for this assignment.
>    2. I will use this Notebook to build the code to scrape the following Wikipedia page,   
>    [https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M](https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M),  
>    in order to obtain the data that is in the table of postal codes and to transform the data into a pandas dataframe like the one shown >below:
>       ![Neighbourhoods|1000x486,50%](pics/wk3pic.jpg) 
>    
>    3. To create the above dataframe:
>       - The dataframe will consist of three columns: __PostalCode, Borough, and Neighborhood__
>       - Only process the cells that have an assigned borough and ignore cells with a borough that is "Not assigned".
>       - More than one neighborhood can exist in one postal code area. For example, in the table on the Wikipedia page, you will notice that M5A is listed twice and has two neighborhoods: Harbourfront and Regent Park. These two rows will be combined into one row with the neighborhoods separated with a comma as shown in row 11 in the above table.
>       - If a cell has a borough but a Not assigned neighborhood, then the neighborhood will be the same as the borough. So for the 9th cell in the table on the Wikipedia page, the value of the Borough and the Neighborhood columns will be Queen's Park.
>       - In the last cell of this notebook, I will use the .shape method to print the number of rows of my dataframe.
>
>    
>    4. I will submit a link to this Notebook on my Github repository. (10 marks)
>
>    5. I will use pandas to scrape the webpage.


<HR>

In [ ]:
# uncomment this line if you haven't completed the Foursquare API lab
!conda install -c conda-forge geopy --yes

In [2]:
# uncomment this line if you haven't completed the Foursquare API lab
!conda install -c conda-forge folium=0.5.0 --yes

Solving environment: ...working... done

# All requested packages already installed.



In [3]:
# Import Libraries to be used

# library to handle data in a vectorized manner
import numpy as np 

# library for data analsysis
import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

# library to handle JSON files
import json 

# convert an address into latitude and longitude values
from geopy.geocoders import Nominatim 

# library to handle requests
import requests 

# tranform JSON file into a pandas dataframe
from pandas.io.json import json_normalize 

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

# map rendering library
import folium 

print('Libraries imported.')

#import lxml.html       as lh #https://lxml.de/lxmlhtml.html

Libraries imported.


In [4]:
# using pandas to scrape the website
url  = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M' 
tables = pd.read_html(url)
tables[0].head()

Postcode           Borough     Neighbourhood
0      M1A      Not assigned      Not assigned
1      M2A      Not assigned      Not assigned
2      M3A        North York         Parkwoods
3      M4A        North York  Victoria Village
4      M5A  Downtown Toronto      Harbourfront

In [5]:
df = tables[0]

# Determine the dimension
df.shape 

(288, 3)

In [6]:
# Determine how many Borough's are 'Not assigned'
print(df['Borough'].value_counts())

Not assigned        77
Etobicoke           45
North York          38
Scarborough         37
Downtown Toronto    37
Central Toronto     17
West Toronto        13
York                 9
East Toronto         7
East York            6
Mississauga          1
Queen's Park         1
Name: Borough, dtype: int64


In [7]:
# All entries where 'Borough' == 'Not Assigned' are absent from new df2
df2 = df[df.Borough != 'Not assigned'] 

In [8]:
# Determine new Dimenion where there are no more Borough's with value 'Not assigned'
df2.shape 

(211, 3)

> ##### <strong> So a drop from 288 to 211. 
    - This means 77 record where Borough = 'Not assigned' were eliminated 

In [9]:
#A check to be sure
print(df2['Borough'].value_counts())

Etobicoke           45
North York          38
Scarborough         37
Downtown Toronto    37
Central Toronto     17
West Toronto        13
York                 9
East Toronto         7
East York            6
Mississauga          1
Queen's Park         1
Name: Borough, dtype: int64


> ##### <strong> Now we look if there are remaining Neigbourhoods that are 'Not assigned'

In [10]:
count_Not_Assigned = df2['Neighbourhood'].str.contains('Not assigned').sum()
if count_Not_Assigned > 0:
    print ("There are {m} Neighbourhoods 'Not_Assigned' ".format(m = count_Not_Assigned))
else: print ("All is good")

There are 1 Neighbourhoods 'Not_Assigned' 


In [11]:
# Find the records where Neighbourhood == 'Not assigned'
df2.loc[df2.Neighbourhood == 'Not assigned']

Postcode       Borough Neighbourhood
8      M7A  Queen's Park  Not assigned

> ##### <strong> Now replace values of 'Not assigned'. Checking total dataframe in case there had been more then 1

In [12]:
for row_index,row in df2.iterrows():
    if df2['Neighbourhood'][row_index]=='Not assigned':
        df2['Neighbourhood'][row_index] = df2['Borough'][row_index]
    else:
        continue

C:\_MyTools\Anaconda\lib\site-packages\IPython\core\interactiveshell.py:3296: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  exec(code_obj, self.user_global_ns, self.user_ns)


In [13]:
df2.loc[df2.Postcode == 'M7A']

Postcode       Borough Neighbourhood
8      M7A  Queen's Park  Queen's Park

> ##### __The swap was succesful__
 - All that we now have to do is to group the neighbourhoods per Postcode

In [14]:
# First some more checking.
df_schoon = df2.reset_index(drop=True)
print(df_schoon.shape)
df_schoon.sort_values(by='Postcode').head(6)

(211, 3)


Postcode      Borough   Neighbourhood
8       M1B  Scarborough           Rouge
9       M1B  Scarborough         Malvern
23      M1C  Scarborough      Port Union
22      M1C  Scarborough      Rouge Hill
21      M1C  Scarborough  Highland Creek
33      M1E  Scarborough       Guildwood

> ##### __Now groupby Postcode and Bouorough__ 
   - to see same postcalcodes with different neighbourhoods.
   - to use apply to join neighbourhoods together when they have the same postalcode and place result back into a dataframe

In [15]:
df_finish = df_schoon.groupby(['Postcode','Borough'])['Neighbourhood'].apply(', '.join).to_frame(name = 'Neighbourhood').reset_index()

In [16]:
#df_finish.to_csv('Tor_Post.csv', sep=';') See results in csv
df_finish.head()

Postcode      Borough                           Neighbourhood
0      M1B  Scarborough                          Rouge, Malvern
1      M1C  Scarborough  Highland Creek, Rouge Hill, Port Union
2      M1E  Scarborough       Guildwood, Morningside, West Hill
3      M1G  Scarborough                                  Woburn
4      M1H  Scarborough                               Cedarbrae

In [17]:
df_finish.shape

(103, 3)

> ##### __We can see that there are 103 records with unique postal codes__

> ##### GitHub-link for this Notebook : [Capstone wk3 - IBM Data Science Clustering Neighbourhoods.ipynb](https://github.com/MarcelVos1966/Capstone-project/blob/master/Capstone%20wk3%20-%20IBM%20Data%20Science%20Clustering%20Neighbourhoods.ipynb)

<HR>

> # __Assignment 2:__
> #### I wil use a csv file that has the geographical coordinates of each postal code: http://cocl.us/Geospatial_data
> - Because of the instability to get the geographical coordinates of the neighborhoods using the Geocoder package <br>
> It wil produce the following dataframe.<br>
> ![Geo Coordinates|708x381](pics/wk3geo.jpg) 
<HR>

In [18]:
# Import csv_file with coordinates
df_geo = pd.read_csv('Geospatial_Coordinates.csv', sep=',')

In [19]:
df_geo.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [20]:
# Determine dimension
df_geo.shape

(103, 3)

> ##### __Unify in name the columns with the 3 alfanumeric code (potcode and Postal Code)__

In [21]:
df_finish.rename(columns={'Postcode': 'Postalcode'}, inplace=True)
df_geo.rename(columns={'Postal Code': 'Postalcode'}, inplace=True)
df_geo.head()

Postalcode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

> ##### __Merge the dataframes to get the final dataframe with all relevant data combined.__
  - merging is done per Postalcode

In [22]:
df_merged_data = pd.merge(df_finish, df_geo, on='Postalcode')
#df_merged_data.to_csv('TorGeo.csv', sep=';')

In [23]:
df_merged_data.head(10)

Postalcode      Borough                                    Neighbourhood  \
0        M1B  Scarborough                                   Rouge, Malvern   
1        M1C  Scarborough           Highland Creek, Rouge Hill, Port Union   
2        M1E  Scarborough                Guildwood, Morningside, West Hill   
3        M1G  Scarborough                                           Woburn   
4        M1H  Scarborough                                        Cedarbrae   
5        M1J  Scarborough                              Scarborough Village   
6        M1K  Scarborough      East Birchmount Park, Ionview, Kennedy Park   
7        M1L  Scarborough                  Clairlea, Golden Mile, Oakridge   
8        M1M  Scarborough  Cliffcrest, Cliffside, Scarborough Village West   
9        M1N  Scarborough                      Birch Cliff, Cliffside West   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476  
5  43.744734 -79.239476  
6  43.727929 -79.262029  
7  43.711112 -79.284577  
8  43.716316 -79.239476  
9  43.692657 -79.264848

<HR>

> # __Assignment 3:__  
> - Explore and cluster the neighborhoods in Toronto.<br>
> - I will zoom in on the district 'Downtown Toronto' and replicate the same analysis as I practiced for the New York City data.

In [24]:
df_neighborhoods = df_merged_data
print(df_neighborhoods.shape)
df_neighborhoods.head()

(103, 5)


Postalcode      Borough                           Neighbourhood   Latitude  \
0        M1B  Scarborough                          Rouge, Malvern  43.806686   
1        M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [25]:
# Use geopy library to get the latitude and longitude values of New York City.
# In order to define an instance of the geocoder, we need to define a user_agent. 
# I will name my agent toronto_explorer, as shown below.
from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

address = 'Toronto'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinates of Toronto are 43.653963, -79.387207.


>#### __Create a map of Toronto with neighborhoods superimposed on top.__

In [26]:
# create map of New York using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, nhood in zip(df_neighborhoods['Latitude'], df_neighborhoods['Longitude'], df_neighborhoods['Borough'], df_neighborhoods['Neighbourhood']):
    label = '{}, {}'.format(nhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=3,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  
    
map_toronto

> __Bescause Github does not show this folium generated map I made a pic of it and imported it__

![Toronto map](pics/Toronto.PNG) 

> ##### Zooming in on __'Downtown Toronto'__

In [27]:
downtown_data = df_neighborhoods[df_neighborhoods['Borough'] == 'Downtown Toronto'].reset_index(drop=True)
downtown_data.head()

Postalcode           Borough                Neighbourhood   Latitude  \
0        M4W  Downtown Toronto                     Rosedale  43.679563   
1        M4X  Downtown Toronto  Cabbagetown, St. James Town  43.667967   
2        M4Y  Downtown Toronto         Church and Wellesley  43.665860   
3        M5A  Downtown Toronto    Harbourfront, Regent Park  43.654260   
4        M5B  Downtown Toronto     Ryerson, Garden District  43.657162   

   Longitude  
0 -79.377529  
1 -79.367675  
2 -79.383160  
3 -79.360636  
4 -79.378937

In [28]:
downtown_data.shape

(18, 5)

> __There are 18 postalcode areas in Downtown Toronto__

In [29]:
address = 'Downtown Toronto'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Downtown Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Downtown Toronto are 43.6541737, -79.3808116451341.


In [30]:
# create map of "Downtown Toronto" using latitude and longitude values
map_downtown = folium.Map(location=[latitude, longitude], zoom_start=13)

# add markers to map
for lat, lng, label in zip(downtown_data['Latitude'], downtown_data['Longitude'], downtown_data['Neighbourhood']):
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_downtown)  
    
map_downtown

> __Bescause Github does not show this folium generated map I made a pic of it and imported it__

![DownTown map](pics/DownTown.PNG) 

> ## __Client_ID data will be deleted from the uploaded file !__

In [31]:
CLIENT_ID     = 'ZBYQMESPW0CO0X2TLE5XPJNA5JEWT1XX1YTEBMSJQT3WVB53' # your Foursquare ID
CLIENT_SECRET = '42ZFBR0DM2CRJQA2LLIXTWT5XATGIVZ1EL2TV4M3VBACNK1D' # your Foursquare Secret
VERSION = '20180604'
LIMIT = 50 #30
print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ZBYQMESPW0CO0X2TLE5XPJNA5JEWT1XX1YTEBMSJQT3WVB53
CLIENT_SECRET:42ZFBR0DM2CRJQA2LLIXTWT5XATGIVZ1EL2TV4M3VBACNK1D


>### Explore the __first__ neighborhood in my dataframe.

Get the neighborhood's name.

In [32]:
downtown_data.loc[0, 'Neighbourhood']

'Rosedale'

In [33]:
neighborhood_latitude  = downtown_data.loc[0, 'Latitude'] # neighborhood latitude value
neighborhood_longitude = downtown_data.loc[0, 'Longitude'] # neighborhood longitude value
neighborhood_name      = downtown_data.loc[0, 'Neighbourhood'] # neighborhood name

print('Latitude and longitude values of {} are {}, {}.'.format(neighborhood_name, 
                                                               neighborhood_latitude, 
                                                               neighborhood_longitude))

Latitude and longitude values of Rosedale are 43.6795626, -79.37752940000001.


> ##### Now, let's get up to the top 100 venues that are in Rosedale within a radius of 500 meters.  
> ##### Step 1 ==> Create the GET request URL. NOt shown to protect client_ID and client_secret

In [34]:
# @Hide
LIMIT  = 100 # limit of number of venues returned by Foursquare API
radius = 500 # define radius
# create URL
url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
    CLIENT_ID, 
    CLIENT_SECRET, 
    VERSION, 
    neighborhood_latitude, 
    neighborhood_longitude, 
    radius, 
    LIMIT)
#url # display URL

> #### Show results

In [35]:
results = requests.get(url).json()
results

{'meta': {'code': 200, 'requestId': '5d3618796f0aa20039f08657'},
 'response': {'headerLocation': 'Rosedale',
  'headerFullLocation': 'Rosedale, Toronto',
  'headerLocationGranularity': 'neighborhood',
  'totalResults': 5,
  'suggestedBounds': {'ne': {'lat': 43.6840626045, 'lng': -79.37131878274371},
   'sw': {'lat': 43.675062595499995, 'lng': -79.38374001725632}},
  'groups': [{'type': 'Recommended Places',
    'name': 'recommended',
    'items': [{'reasons': {'count': 0,
       'items': [{'summary': 'This spot is popular',
         'type': 'general',
         'reasonName': 'globalInteractionReason'}]},
      'venue': {'id': '4bae2150f964a520df873be3',
       'name': 'Mooredale House',
       'location': {'address': '146 Crescent Rd.',
        'crossStreet': 'btwn. Lamport Ave. and Mt. Pleasant Rd.',
        'lat': 43.678630645646535,
        'lng': -79.38009142511322,
        'labeledLatLngs': [{'label': 'display',
          'lat': 43.678630645646535,
          'lng': -79.380091425113

> From the Foursquare lab in the previous module, I know that all the information is in the items key. 
Before I proceed, I'll borrow the get_category_type function from the Foursquare lab.


In [36]:
# function that extracts the category of the venue
def get_category_type(row):
    try:
        categories_list = row['categories']
    except:
        categories_list = row['venue.categories']
        
    if len(categories_list) == 0:
        return None
    else:
        return categories_list[0]['name']

In [37]:
import json                               # library to handle JSON files
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

In [38]:
venues = results['response']['groups'][0]['items']
    
nearby_venues = json_normalize(venues) # flatten JSON

# filter columns
filtered_columns = ['venue.name', 'venue.categories', 'venue.location.lat', 'venue.location.lng']
nearby_venues =nearby_venues.loc[:, filtered_columns]

# filter the category for each row
nearby_venues['venue.categories'] = nearby_venues.apply(get_category_type, axis=1)

# clean columns
nearby_venues.columns = [col.split(".")[-1] for col in nearby_venues.columns]

nearby_venues.head()

name  categories        lat        lng
0       Mooredale House    Building  43.678631 -79.380091
1         Rosedale Park  Playground  43.682328 -79.378934
2          Whitney Park        Park  43.682036 -79.373788
3  Alex Murray Parkette        Park  43.678300 -79.382773
4        Milkman's Lane       Trail  43.676352 -79.373842

> And how many venues were returned by Foursquare?

In [39]:
print('{} venues were returned by Foursquare.'.format(nearby_venues.shape[0]))

5 venues were returned by Foursquare.


> __Apparently not much happening around Rosedale which is mostly a Park District__   
> So not so surprising after all

# Explore Neighborhoods in Downtown Toronto

In [40]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

> #### Now the code to run the above function on each neighborhood and create a new dataframe called *downtown_venues*.

In [41]:
downtown_venues = getNearbyVenues(names      = downtown_data['Neighbourhood'],
                                  latitudes  = downtown_data['Latitude'],
                                  longitudes = downtown_data['Longitude']
                                  )


Rosedale
Cabbagetown, St. James Town
Church and Wellesley
Harbourfront, Regent Park
Ryerson, Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide, King, Richmond
Harbourfront East, Toronto Islands, Union Station
Design Exchange, Toronto Dominion Centre
Commerce Court, Victoria Hotel
Harbord, University of Toronto
Chinatown, Grange Park, Kensington Market
CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place, Underground city
Christie


> ##### __Check the size of the resulting dataframe__

In [42]:
print(downtown_venues.shape)
downtown_venues.head()

(1290, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0     Rosedale              43.679563              -79.377529   
1     Rosedale              43.679563              -79.377529   
2     Rosedale              43.679563              -79.377529   
3     Rosedale              43.679563              -79.377529   
4     Rosedale              43.679563              -79.377529   

                  Venue  Venue Latitude  Venue Longitude Venue Category  
0       Mooredale House       43.678631       -79.380091       Building  
1         Rosedale Park       43.682328       -79.378934     Playground  
2          Whitney Park       43.682036       -79.373788           Park  
3  Alex Murray Parkette       43.678300       -79.382773           Park  
4        Milkman's Lane       43.676352       -79.373842          Trail

> ##### __Let's check how many venues were returned for each neighborhood__


In [43]:
downtown_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                                              100   
Berczy Park                                                            57   
CN Tower, Bathurst Quay, Island airport, Harbou...                     16   
Cabbagetown, St. James Town                                            44   
Central Bay Street                                                     84   
Chinatown, Grange Park, Kensington Market                             100   
Christie                                                               15   
Church and Wellesley                                                   89   
Commerce Court, Victoria Hotel                                        100   
Design Exchange, Toronto Dominion Centre                              100   
First Canadian Place, Underground city                                100   
Harbord, University of Toronto                                         34   
Harbourfront East, Toronto Islands, Union Station                     100   
Harbourfront, Regent Park                                              51   
Rosedale                                                                5   
Ryerson, Garden District                                              100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        95   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide, King, Richmond                                               100   
Berczy Park                                                             57   
CN Tower, Bathurst Quay, Island airport, Harbou...                      16   
Cabbagetown, St. James Town                                             44   
Central Bay Street                                                      84   
Chinatown, Grange Park, Kensington Market                              100   
Christie                                                                15   
Church and Wellesley                                                    89   
Commerce Court, Victoria Hotel                                         100   
Design Exchange, Toronto Dominion Centre                               100   
First Canadian Place, Underground city                                 100   
Harbord, University of Toronto                                          34   
Harbourfront East, Toronto Islands, Union Station                      100   
Harbourfront, Regent Park                                               51   
Rosedale                                                                 5   
Ryerson, Garden District                                               100   
St. James Town                                                         100   
Stn A PO Boxes 25 The Esplanade                                         95   

                                                    Venue  Venue Latitude  \
Neighborhood                                                                
Adelaide, King, Richmond                              100             100   
Berczy Park                                            57              57   
CN Tower, Bathurst Quay, Island airport, Harbou...     16              16   
Cabbagetown, St. James Town                            44              44   
Central Bay Street                                     84              84   
Chinatown, Grange Park, Kensington Market             100             100   
Christie                                               15              15   
Church and Wellesley                                   89              89   
Commerce Court, Victoria Hotel                        100             100   
Design Exchange, Toronto Dominion Centre              100             100   
First Canadian Place, Unde

> __Find out how many unique categories can be curated from all the returned venues__

In [44]:
print('There are {} uniques categories.'.format(len(downtown_venues['Venue Category'].unique())))

There are 204 uniques categories.


In [45]:
# one hot encoding
downtown_onehot = pd.get_dummies(downtown_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
downtown_onehot['Neighborhood'] = downtown_venues['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [downtown_onehot.columns[-1]] + list(downtown_onehot.columns[:-1])
downtown_onehot = downtown_onehot[fixed_columns]

downtown_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  \
0             0         0            0           0                    0   
1             0         0            0           0                    0   
2             0         0            0           0                    0   
3             0         0            0           0                    0   
4             0         0            0           0                    0   

   Asian Restaurant  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0                 0          0           0           0       0     0    0   
1                 0          0           0           0       0     0    0   
2                 0          0           0           0       0     0    0   
3                 0          0           0           0       0     0    0   
4                 0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Beer Bar  Beer Store  \
0                 0                   0      0         0           0   
1                 0                   0      0         0           0   
2                 0                   0      0         0           0   
3                 0                   0      0         0           0   
4                 0                   0      0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         1   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Butcher  Café  Camera Store  \
0             0              0        0     0             0   
1             0              0        0     0             0   
2             0              0        0     0             0   
3             0              0        0     0             0   
4             0              0        0     0             0   

   Caribbean Restaurant  Cheese Shop  Chinese Restaurant  Chocolate Shop  \
0                     0            0                   0               0   
1                     0            0                   0               0   
2                     0            0                   0               0   
3                     0            0                   0               0   
4                     0            0                   0       

In [46]:
downtown_onehot.shape

(1290, 204)

> __Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category__

In [47]:
downtown_grouped = downtown_onehot.groupby('Neighborhood').mean().reset_index()
downtown_grouped[0:2]

Neighborhood  Yoga Studio  Afghan Restaurant  Airport  \
0  Adelaide, King, Richmond          0.0                0.0      0.0   
1               Berczy Park          0.0                0.0      0.0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                 0.0           0.0             0.0              0.0   
1                 0.0           0.0             0.0              0.0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0               0.0                 0.03           0.0       0.0     0.010000   
1               0.0                 0.00           0.0       0.0     0.017544   

   Art Museum  Arts & Crafts Store  Asian Restaurant  BBQ Joint  Baby Store  \
0        0.01                  0.0              0.03        0.0         0.0   
1        0.00                  0.0              0.00        0.0         0.0   

   Bagel Shop    Bakery  Bank   Bar  Baseball Stadium  Basketball Stadium  \
0    0.000000  0.020000   0.0  0.04               0.0            0.000000   
1    0.017544  0.035088   0.0  0.00               0.0            0.017544   

      Beach  Beer Bar  Beer Store  Belgian Restaurant    Bistro  \
0  0.000000  0.000000         0.0            0.000000  0.000000   
1  0.017544  0.035088         0.0            0.017544  0.017544   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0            0.0       0.01       0.0                  0.01        0.020000   
1            0.0       0.00       0.0                  0.00        0.017544   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place   Butcher  \
0      0.0              0.0      0.01          0.03           0.01  0.000000   
1      0.0              0.0      0.00          0.00           0.00  0.017544   

       Café  Camera Store  Caribbean Restaurant  Cheese Shop  \
0  0.050000           0.0                   0.0     0.000000   
1  0.035088           0.0                   0.0     0.035088   

   Chinese Restaurant  Chocolate Shop  Church  Clothing Store  Cocktail Bar  \
0                 0.0             0.0     0.0        0.010000      0.000000   
1                 0.0             0.0     0.0        0.017544      0.052632   

   Coffee Shop  College Arts Building  College Gym  College Rec Center  \
0     0.070000                    0.0          0.0                 0.0   
1     0.105263                    0.0          0.0                 0.0   

   Colombian Restaurant  Comfort Food Restaurant  Comic Shop  Concert Hall  \
0                  0.01                 0.000000         0.0      0.020000   
1                  0.00                 0.017544         0.0      0.017544   

   Convenience Store  Cosmetics Shop  Creperie  Dance Studio  Deli / Bodega  \
0                0.0        0.030000  0.000000          0.01           0.01   
1                0.0        0.017544  0.017544          0.00           0.00   

   Department Store  Dessert Shop  Dim Sum Restaurant     Diner  \
0              0.01           0.0                 0.0  0.000000   
1              0.00           0.0                 0.0  0.017544   

   Discount Store  Dive Bar  Dog Run  Doner Restaurant  Donut Shop  \
0             0.0       0.0      0.0               0.0         0.0   
1             0.0       0.0      0.0               0.0         0.0   

   Dumpling Restaurant  Eastern European Restaurant  Electronics Store  \
0                  0.0                     0.000000               0.01   
1                  0.0                     0.017544               0.00   

   Ethiopian Restaurant  Event Space  Falafel Restaurant  Farmers Market  \
0                   0.0          0.0                 0.0        0.000000   
1                   0.0          0.0                 0.0        0.035088   

   Fast Food Restaurant  Filipino Restaurant  Fish & Chips Shop  Fish Market  \
0                   0.0                  0.0                0.0     0.000000   
1                   0.0           

> ##### Let's confirm the new size

In [48]:
downtown_grouped.shape

(18, 204)

> __Let's print each neighborhood along with the top 5 most common venues__

In [49]:
num_top_venues = 5

for hood in downtown_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = downtown_grouped[downtown_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide, King, Richmond----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2              Bar  0.04
3  Thai Restaurant  0.04
4       Steakhouse  0.04


----Berczy Park----
                venue  freq
0         Coffee Shop  0.11
1        Cocktail Bar  0.05
2  Seafood Restaurant  0.04
3      Farmers Market  0.04
4            Beer Bar  0.04


----CN Tower, Bathurst Quay, Island airport, Harbourfront West, King and Spadina, Railway Lands, South Niagara----
              venue  freq
0   Airport Service  0.19
1    Airport Lounge  0.12
2  Airport Terminal  0.12
3   Harbor / Marina  0.06
4               Bar  0.06


----Cabbagetown, St. James Town----
         venue  freq
0   Restaurant  0.07
1       Bakery  0.07
2  Coffee Shop  0.07
3       Market  0.05
4  Pizza Place  0.05


----Central Bay Street----
                venue  freq
0         Coffee Shop  0.14
1                Café  0.05
2      Ice Cream Shop  0.05
3  Italian Restaurant  0.05
4      Sandwich Place

> ### Let's put that into a pandas dataframe
> __First, A function to sort the venues in descending order__

In [50]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

> __Create the new dataframe and display the top 10 venues for each neighborhood.__

In [51]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = downtown_grouped['Neighborhood']

for ind in np.arange(downtown_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(downtown_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2  CN Tower, Bathurst Quay, Island airport, Harbo...       Airport Service   
3                        Cabbagetown, St. James Town            Restaurant   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café            Steakhouse       Thai Restaurant   
1          Cocktail Bar                  Café           Cheese Shop   
2      Airport Terminal        Airport Lounge       Harbor / Marina   
3           Coffee Shop                Bakery           Pizza Place   
4        Ice Cream Shop    Italian Restaurant                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Bar          Burger Joint                   Gym   
1              Beer Bar                Bakery    Seafood Restaurant   
2         Boat or Ferry                   Bar      Sculpture Garden   
3                  Café                Market    Italian Restaurant   
4        Sandwich Place          Burger Joint    Chinese Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                 Hotel      Asian Restaurant    American Restaurant  
1            Steakhouse        Farmers Market                 Lounge  
2           Coffee Shop              Boutique           Airport Gate  
3                   Pub         Jewelry Store           Liquor Store  
4                   Spa      Sushi Restaurant    Japanese Restaurant

> # __Cluster Neighborhoods__
 - Run k-means to cluster the neighborhood into 5 clusters.


In [52]:
# set number of clusters
kclusters = 5

downtown_grouped_clustering = downtown_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(downtown_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 3, 0, 0, 2, 4, 0, 0, 0])

> __Create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.__

In [53]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

downtown_merged = downtown_data
downtown_merged

Postalcode           Borough  \
0         M4W  Downtown Toronto   
1         M4X  Downtown Toronto   
2         M4Y  Downtown Toronto   
3         M5A  Downtown Toronto   
4         M5B  Downtown Toronto   
5         M5C  Downtown Toronto   
6         M5E  Downtown Toronto   
7         M5G  Downtown Toronto   
8         M5H  Downtown Toronto   
9         M5J  Downtown Toronto   
10        M5K  Downtown Toronto   
11        M5L  Downtown Toronto   
12        M5S  Downtown Toronto   
13        M5T  Downtown Toronto   
14        M5V  Downtown Toronto   
15        M5W  Downtown Toronto   
16        M5X  Downtown Toronto   
17        M6G  Downtown Toronto   

                                        Neighbourhood   Latitude  Longitude  
0                                            Rosedale  43.679563 -79.377529  
1                         Cabbagetown, St. James Town  43.667967 -79.367675  
2                                Church and Wellesley  43.665860 -79.383160  
3                           Harbourfront, Regent Park  43.654260 -79.360636  
4                            Ryerson, Garden District  43.657162 -79.378937  
5                                      St. James Town  43.651494 -79.375418  
6                                         Berczy Park  43.644771 -79.373306  
7                                  Central Bay Street  43.657952 -79.387383  
8                            Adelaide, King, Richmond  43.650571 -79.384568  
9   Harbourfront East, Toronto Islands, Union Station  43.640816 -79.381752  
10           Design Exchange, Toronto Dominion Centre  43.647177 -79.381576  
11                     Commerce Court, Victoria Hotel  43.648198 -79.379817  
12                     Harbord, University of Toronto  43.662696 -79.400049  
13          Chinatown, Grange Park, Kensington Market  43.653206 -79.400049  
14  CN Tower, Bathurst Quay, Island airport, Harbo...  43.628947 -79.394420  
15                    Stn A PO Boxes 25 The Esplanade  43.646435 -79.374846  
16             First Canadian Place, Underground city  43.648429 -79.382280  
17                                           Christie  43.669542 -79.422564

In [54]:
# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
downtown_merged = downtown_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighbourhood')
downtown_merged.head() # check the last columns!

Postalcode           Borough                Neighbourhood   Latitude  \
0        M4W  Downtown Toronto                     Rosedale  43.679563   
1        M4X  Downtown Toronto  Cabbagetown, St. James Town  43.667967   
2        M4Y  Downtown Toronto         Church and Wellesley  43.665860   
3        M5A  Downtown Toronto    Harbourfront, Regent Park  43.654260   
4        M5B  Downtown Toronto     Ryerson, Garden District  43.657162   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.377529               1                  Park            Playground   
1 -79.367675               0            Restaurant           Coffee Shop   
2 -79.383160               0           Coffee Shop   Japanese Restaurant   
3 -79.360636               0           Coffee Shop                Bakery   
4 -79.378937               0           Coffee Shop        Clothing Store   

  3rd Most Common Venue 4th Most Common Venue      5th Most Common Venue  \
0                 Trail              Building               Dessert Shop   
1                Bakery           Pizza Place                       Café   
2               Gay Bar      Sushi Restaurant                 Restaurant   
3                   Pub                  Café                       Park   
4        Cosmetics Shop  Fast Food Restaurant  Middle Eastern Restaurant   

         6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0  Eastern European Restaurant   Dumpling Restaurant            Donut Shop   
1                       Market    Italian Restaurant                   Pub   
2                        Hotel             Gastropub                   Pub   
3                      Theater        Breakfast Spot    Mexican Restaurant   
4                         Café                 Diner        Lingerie Store   

  9th Most Common Venue 10th Most Common Venue  
0      Doner Restaurant                Dog Run  
1         Jewelry Store           Liquor Store  
2  Fast Food Restaurant           Dance Studio  
3            Restaurant   Gym / Fitness Center  
4        Ice Cream Shop        Bubble Tea Shop

# __Visualize__ resulting clusters by __creating a map__

In [62]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(downtown_merged['Latitude'], downtown_merged['Longitude'], downtown_merged['Neighbourhood'], downtown_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

> __Bescause Github does not show this folium generated map I made a pic of it and imported it__

![DownTown map with clusters](pics/DowntownClusters.PNG) 

> __Examine each cluster and determine the discriminating venue categories that distinguish each cluster.__    
Based on the defining categories, I will then assign a name to each cluster. 

#### __Cluster 1__

In [56]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 0, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1   Downtown Toronto               0            Restaurant   
2   Downtown Toronto               0           Coffee Shop   
3   Downtown Toronto               0           Coffee Shop   
4   Downtown Toronto               0           Coffee Shop   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0           Coffee Shop   
8   Downtown Toronto               0           Coffee Shop   
9   Downtown Toronto               0           Coffee Shop   
10  Downtown Toronto               0           Coffee Shop   
11  Downtown Toronto               0           Coffee Shop   
15  Downtown Toronto               0           Coffee Shop   
16  Downtown Toronto               0           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
1            Coffee Shop                Bakery           Pizza Place   
2    Japanese Restaurant               Gay Bar      Sushi Restaurant   
3                 Bakery                   Pub                  Café   
4         Clothing Store        Cosmetics Shop  Fast Food Restaurant   
5                   Café            Restaurant                 Hotel   
6           Cocktail Bar                  Café           Cheese Shop   
7         Ice Cream Shop    Italian Restaurant                  Café   
8                   Café            Steakhouse       Thai Restaurant   
9                  Hotel              Aquarium                  Café   
10                  Café                 Hotel            Restaurant   
11                 Hotel                  Café   American Restaurant   
15            Restaurant                  Café              Beer Bar   
16                  Café            Restaurant                 Hotel   

        5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
1                        Café                Market    Italian Restaurant   
2                  Restaurant                 Hotel             Gastropub   
3                        Park               Theater        Breakfast Spot   
4   Middle Eastern Restaurant                  Café                 Diner   
5          Italian Restaurant              Beer Bar             Gastropub   
6                    Beer Bar                Bakery    Seafood Restaurant   
7              Sandwich Place          Burger Joint    Chinese Restaurant   
8                         Bar          Burger Joint                   Gym   
9          Italian Restaurant                Bakery               Brewery   
10         Italian Restaurant                   Gym         Deli / Bodega   
11                 Restaurant                Bakery            Steakhouse   
15         Seafood Restaurant  Fast Food Restaurant          Cocktail Bar   
16                 Steakhouse                   Bar    Seafood Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
1                    Pub         Jewelry Store           Liquor Store  
2                    Pub  Fast Food Restaurant           Dance Studio  
3     Mexican Restaurant            Restaurant   Gym / Fitness Center  
4         Lingerie Store        Ice Cream Shop        Bubble Tea Shop  
5         Breakfast Spot        Cosmetics Shop           Cocktail Bar  
6             Steakhouse        Farmers Market                 Lounge  
7                    Spa      Sushi Restaurant    Japanese Restaurant  
8                  Hotel      Asian Restaurant    American Restaurant  
9            Pizza Place        Scenic Lookout    Sporting Goods Shop  
10                   Bar             Gastropub                 Bakery  
11                   Gym         Deli / Bodega              Gastropub  
15                 Hotel           Cheese Shop            Art Gallery  
16                   Gym             Gastropub          Deli / Bodega

#### __Cluster 2__

In [57]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 1, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0  Downtown Toronto               1                  Park   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0            Playground                 Trail              Building   

  5th Most Common Venue        6th Most Common Venue 7th Most Common Venue  \
0          Dessert Shop  Eastern European Restaurant   Dumpling Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Donut Shop      Doner Restaurant                Dog Run

#### __Cluster 3__

In [58]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 2, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
12  Downtown Toronto               2                  Café   
13  Downtown Toronto               2                  Café   

            2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
12                     Restaurant             Bookstore   Japanese Restaurant   
13  Vegetarian / Vegan Restaurant    Chinese Restaurant                   Bar   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
12                   Bar                Bakery    Chinese Restaurant   
13    Mexican Restaurant   Dumpling Restaurant                Bakery   

    8th Most Common Venue 9th Most Common Venue   10th Most Common Venue  
12         Sandwich Place                   Pub  Comfort Food Restaurant  
13  Vietnamese Restaurant           Coffee Shop             Cocktail Bar

#### __Cluster 4__

In [59]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 3, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
14  Downtown Toronto               3       Airport Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
14      Airport Terminal        Airport Lounge       Harbor / Marina   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
14         Boat or Ferry                   Bar      Sculpture Garden   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
14           Coffee Shop              Boutique           Airport Gate

#### __Cluster 5__

In [60]:
downtown_merged.loc[downtown_merged['Cluster Labels'] == 4, downtown_merged.columns[[1] + list(range(5, downtown_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
17  Downtown Toronto               4         Grocery Store   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
17                  Café                  Park             Nightclub   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
17                 Diner    Italian Restaurant            Baby Store   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
17            Restaurant           Coffee Shop      Convenience Store

#### __Check if there is a Cluster 6__

># Assigning names to each cluster
> -  Cluster 1: Center district
> -  Cluster 2: Parks district
> -  Cluster 3: Asian district (Includes University)
> -  Cluster 4: Airport-Harbour district
> -  Cluster 5: Public recreational district (Christie pits)

<HR>

> # __End Notebook__
>
> ![Geo Coordinates|708x381](pics/end.png) 

<HR>